File: tests/test_output_name.py
"""
Tests for _output_name — what the built PDFs are called.

The filename is the first thing a recruiter sees, before the document
opens, and it is also the most fragile part of the deliverable: it
passes through mail clients, HR portals and applicant-tracking systems
that are far less careful with bytes than the PDF's own metadata. So
the stem is reduced to ASCII, and these tests pin down exactly what
that reduction does to a name.

TWO FAILURES ARE WORSE THAN THE OTHERS, AND BOTH ARE SILENT
───────────────────────────────────────────────────────────
Losing a letter. Stripping "José" to "Jos" produces a filename that
looks fine, builds fine, and is somebody else's name. Every accented
character must fold to its base letter, and the atomic ones NFKD
cannot decompose (ø, ß, æ, ł …) must be transliterated by hand or they
vanish the same way — see TestFolding.

Producing an unwritable name. `name.first` is free text in a YAML file;
nothing stops a colon, a slash or a newline from ending up there, and
on Windows any of them makes the build fail at the last step, after
everything else succeeded. See TestFilesystemSafety.

The third class, TestDegenerateInput, covers what happens when there is
no usable name at all — a missing field, an empty string, a name with
no ASCII reading. The build must still have somewhere to write.
"""

import json
import sys
import unittest
from pathlib import Path

ROOT = Path(__file__).parent.parent
sys.path.insert(0, str(ROOT / "build"))

import _output_name as on  # noqa: E402


def stem(first, last, variant="resume"):
    """Shorthand: the stem for a {first, last} name."""
    return on.stem({"first": first, "last": last}, variant)


class TestTheConvention(unittest.TestCase):
    """The shape the user asked for: First_Last_Resume."""

    def test_resume_stem(self):
        self.assertEqual(stem("Gaius", "Caesar"), "Gaius_Caesar_Resume")

    def test_cover_letter_stem(self):
        self.assertEqual(stem("Gaius", "Caesar", "letter"),
                         "Gaius_Caesar_Cover_Letter")

    def test_color_takes_the_bare_stem(self):
        # The color PDF is the one that gets attached to applications,
        # so it gets the clean name; only grayscale is suffixed.
        self.assertEqual(on.color_pdf("dist", "Gaius_Caesar_Resume").name,
                         "Gaius_Caesar_Resume.pdf")
        self.assertEqual(on.grayscale_pdf("dist", "Gaius_Caesar_Resume").name,
                         "Gaius_Caesar_Resume_Grayscale.pdf")

    def test_both_variants_share_one_stem(self):
        # Not decoration: the app pairs them by stem, and pruneStale
        # in _output_name.js keeps whichever two the build produced.
        s = stem("Gaius", "Caesar")
        color = on.color_pdf("dist", s).name
        gray = on.grayscale_pdf("dist", s).name
        self.assertTrue(gray.startswith(color[:-len(".pdf")]))

    def test_unknown_variant_is_refused(self):
        with self.assertRaises(ValueError):
            on.stem({"first": "A", "last": "B"}, "invoice")


class TestFolding(unittest.TestCase):
    """Every letter survives, in some ASCII form."""

    def test_accents_fold_rather_than_strip(self):
        self.assertEqual(stem("José", "Álvarez"), "Jose_Alvarez_Resume")
        self.assertEqual(stem("Zoë", "Brontë"), "Zoe_Bronte_Resume")
        self.assertEqual(stem("François", "Noël"), "Francois_Noel_Resume")

    def test_letters_nfkd_cannot_decompose_are_transliterated(self):
        # These are atomic code points, not base + combining mark.
        # Without the explicit table they would be dropped entirely,
        # which is the silent letter-loss this module exists to avoid.
        self.assertEqual(stem("Søren", "Kierkegaard"), "Soren_Kierkegaard_Resume")
        self.assertEqual(stem("Lech", "Wałęsa"), "Lech_Walesa_Resume")
        self.assertEqual(stem("Strauß", "Hoffmann"), "Strauss_Hoffmann_Resume")
        self.assertEqual(stem("Æthelred", "Þorsteinn"), "Aethelred_Thorsteinn_Resume")

    def test_no_accented_character_disappears(self):
        # A property test over the Latin-1 supplement, which is where
        # European names live. Every letter must contribute at least
        # one ASCII character to the result.
        for code in range(0x00C0, 0x0180):
            ch = chr(code)
            if not ch.isalpha():
                continue
            folded = on.slug_part(f"A{ch}B")
            self.assertTrue(
                len(folded) >= 3,
                f"U+{code:04X} ({ch!r}) vanished: {folded!r}")

    def test_case_is_preserved(self):
        # The name is the name; this is not a slug for a URL.
        self.assertEqual(stem("Gaius", "IULIUS"), "Gaius_IULIUS_Resume")


class TestPunctuationInNames(unittest.TestCase):
    """Real surnames are not [A-Za-z]+."""

    def test_apostrophes_are_dropped_not_separated(self):
        # O_Brien reads like two names. OBrien reads like one.
        self.assertEqual(stem("Seamus", "O'Brien"), "Seamus_OBrien_Resume")
        self.assertEqual(stem("Seamus", "O’Brien"), "Seamus_OBrien_Resume")

    def test_hyphens_and_spaces_become_the_separator(self):
        self.assertEqual(stem("Anne-Marie", "Smith Jones"),
                         "Anne_Marie_Smith_Jones_Resume")

    def test_particles_survive(self):
        self.assertEqual(stem("Ludwig", "van der Berg"),
                         "Ludwig_van_der_Berg_Resume")

    def test_runs_collapse_and_edges_are_trimmed(self):
        self.assertEqual(on.slug_part("  --Gaius--  "), "Gaius")
        self.assertEqual(on.slug_part("A....B"), "A_B")

    def test_a_name_cannot_impersonate_the_variant_suffix(self):
        # GRAYSCALE_SUFFIX is '_Grayscale' — the same separator the name
        # parts use. That is only safe because DOC_SUFFIX always ends
        # the stem, so the marker can never be mistaken for part of a
        # name, however the name is spelled.
        for surname in ("Grayscale", "Smith-Jones", "Grayscale Jones"):
            with self.subTest(surname=surname):
                s = stem("Anne", surname)
                self.assertTrue(s.endswith("_Resume"),
                                f"{s!r} does not end with the document suffix")
                self.assertFalse(s.endswith(on.GRAYSCALE_SUFFIX),
                                 f"{s!r} looks like a grayscale filename")


class TestFilesystemSafety(unittest.TestCase):
    """Nothing the YAML can hold may produce an unwritable name."""

    def test_windows_forbidden_characters_never_survive(self):
        for ch in '\\/:*?"<>|':
            with self.subTest(ch=ch):
                self.assertNotIn(ch, stem("A" + ch + "B", "C"))

    def test_control_characters_and_newlines_never_survive(self):
        result = stem("Gaius\nIulius\tCaesar\x00", "Rex")
        for ch in result:
            self.assertTrue(ch.isprintable(), f"{ch!r} in {result!r}")
        self.assertNotIn(" ", result)

    def test_path_traversal_cannot_be_expressed(self):
        # `name.first` is free text. If a stem could contain a
        # separator, the build would write outside dist/.
        self.assertNotIn("/", stem("../../etc", "passwd"))
        self.assertNotIn("\\", stem("..\\..\\Windows", "System32"))
        self.assertNotIn("..", stem("..", ".."))

    def test_long_names_are_capped_per_part(self):
        # Windows' 260-character path limit is the real constraint.
        long_stem = stem("A" * 200, "B" * 200)
        self.assertLessEqual(len(long_stem), 2 * on.MAX_PART + 32)

    def test_the_cap_does_not_leave_a_trailing_separator(self):
        # Cutting mid-run would otherwise yield "Foo_..._" + "_Resume".
        raw = ("x" * (on.MAX_PART - 1)) + "  yyy"
        self.assertFalse(on.slug_part(raw).endswith("_"))
        self.assertNotIn("__", stem(raw, "Caesar"))


class TestDegenerateInput(unittest.TestCase):
    """There is always somewhere to write."""

    def test_missing_last_name_drops_out_cleanly(self):
        self.assertEqual(on.stem({"first": "Prince"}, "resume"),
                         "Prince_Resume")
        self.assertNotIn("__", on.stem({"first": "Prince"}, "resume"))

    def test_no_name_at_all_falls_back_to_the_document(self):
        self.assertEqual(on.stem({}, "resume"), "Resume")
        self.assertEqual(on.stem({}, "letter"), "Cover_Letter")
        self.assertEqual(on.stem(None, "resume"), "Resume")

    def test_empty_and_whitespace_names_are_treated_as_absent(self):
        self.assertEqual(stem("", ""), "Resume")
        self.assertEqual(stem("   ", "\t"), "Resume")

    def test_non_string_name_fields_do_not_crash(self):
        # YAML 1.2 core typing means a bare `last: 1984` is an int.
        self.assertEqual(on.stem({"first": "Gaius", "last": 42}, "resume"),
                         "Gaius_Resume")

    def test_a_name_with_no_ascii_reading_keeps_its_characters(self):
        # Folding a CJK name to nothing and shipping "Resume.pdf" is
        # not an acceptable answer for the person whose name it is.
        result = stem("李", "世民")
        self.assertIn("李", result)
        self.assertTrue(result.endswith("_Resume"))


class TestMixedScriptNames(unittest.TestCase):
    """A part that would lose letters to folding is kept whole.

    The ASCII pass used to keep only the Latin half: "Zoë 山田" became
    "Zoe", "Petrov-Смирнов" became "Petrov" — a file named after half
    a person.
    """

    def test_latin_and_cjk_in_one_part(self):
        self.assertEqual(on.slug_part("Zoë 山田"), "Zoë_山田")

    def test_latin_and_cyrillic_in_one_part(self):
        self.assertEqual(on.slug_part("Petrov-Смирнов"), "Petrov-Смирнов")

    def test_the_full_stem(self):
        self.assertEqual(
            on.stem({"first": "Zoë 山田", "last": "Petrov-Смирнов"}, "resume"),
            "Zoë_山田_Petrov-Смирнов_Resume")

    def test_pure_latin_parts_are_still_folded(self):
        self.assertEqual(on.slug_part("Zoë"), "Zoe")
        self.assertEqual(on.stem({"first": "Zoë", "last": "山田"}, "resume"),
                         "Zoe_山田_Resume")

    def test_mixed_script_stays_filesystem_safe(self):
        got = on.slug_part('Zoë/山田: "x"*?<>|\x07')
        for ch in '\\/:*?"<>|\x07':
            self.assertNotIn(ch, got)
        self.assertIn("山田", got)

    def test_mixed_script_is_capped(self):
        got = on.slug_part("Zoë " + "山" * 200)
        self.assertLessEqual(len(got), on.MAX_PART)

    def test_symbols_alone_do_not_trigger_the_fallback(self):
        """Only letters count: an emoji or a dash is not a lost letter."""
        self.assertEqual(on.slug_part("Ann – Lee ★"), "Ann_Lee")


class TestMetadataRoundTrip(unittest.TestCase):
    """The stem travels to Node through the build metadata."""

    def test_stem_is_read_back_from_a_metadata_file(self):
        import tempfile
        with tempfile.TemporaryDirectory() as tmp:
            meta = Path(tmp) / "pdf_meta.json"
            meta.write_text(json.dumps({"output_stem": "Gaius_Caesar_Resume"}),
                            encoding="utf-8")
            self.assertEqual(on.stem_from_meta(meta, "resume"),
                             "Gaius_Caesar_Resume")

    def test_a_missing_metadata_file_yields_the_document_suffix(self):
        # The bootstrap case: nothing has been built, so the fallback
        # only ever names a file that is correctly reported as absent.
        self.assertEqual(on.stem_from_meta(ROOT / "nope.json", "resume"),
                         "Resume")
        self.assertEqual(on.stem_from_meta(ROOT / "nope.json", "letter"),
                         "Cover_Letter")

    def test_a_corrupt_metadata_file_does_not_raise(self):
        import tempfile
        with tempfile.TemporaryDirectory() as tmp:
            meta = Path(tmp) / "pdf_meta.json"
            meta.write_text("{ not json", encoding="utf-8")
            self.assertEqual(on.stem_from_meta(meta, "resume"), "Resume")

    def test_a_metadata_file_without_the_field_does_not_raise(self):
        # An older build wrote this file before output_stem existed.
        import tempfile
        with tempfile.TemporaryDirectory() as tmp:
            meta = Path(tmp) / "pdf_meta.json"
            meta.write_text(json.dumps({"title": "x"}), encoding="utf-8")
            self.assertEqual(on.stem_from_meta(meta, "resume"), "Resume")


class TestAgainstTheRealData(unittest.TestCase):
    """The names in data/ produce usable filenames."""

    def test_every_data_file_yields_a_writable_stem(self):
        import build  # noqa: PLC0415 — deliberately late; needs sys.path
        data_dir = ROOT / "data"
        if not data_dir.is_dir():
            self.skipTest("no data/ directory in this checkout")

        seen = 0
        for path in sorted(data_dir.glob("*.yml")):
            doc = build._yaml_loader.load(path.read_text(encoding="utf-8"))
            if not isinstance(doc, dict) or "name" not in doc:
                continue
            seen += 1
            variant = "letter" if "letter" in doc else "resume"
            result = on.stem(doc["name"], variant)
            with self.subTest(file=path.name):
                self.assertTrue(result, "empty stem")
                for ch in '\\/:*?"<>|':
                    self.assertNotIn(ch, result)
                self.assertTrue(
                    result.endswith(on.DOC_SUFFIX[variant]),
                    f"{result!r} does not end with the document suffix")
        # Guard against this passing by iterating over nothing — the
        # exact failure mode that let a repaired test skip forever.
        self.assertGreater(seen, 0, "no named data files were checked")


if __name__ == "__main__":
    unittest.main()

File: tests/test_output_name.js
/**
 * test_output_name.js — the Node half of the output-naming contract.
 *
 * The PDFs are named after you: Gaius_Caesar_Resume.pdf. That name is
 * derived from YAML, which only Python in this project can read, so
 * Node never computes it — build.py writes `output_stem` into
 * dist/pdf_meta.json and this module reads it back. Everything here
 * tests that one-way trip and what happens when it fails, which it
 * will: the app polls for tray status before anything is built, and
 * an interrupted build can leave a half-written metadata file.
 *
 * WHAT IS ACTUALLY AT RISK HERE
 * ─────────────────────────────
 * pruneStale() deletes files. It exists because the set of filenames a
 * build occupies moves when your name does, and a stranded PDF in
 * dist/ is not clutter — it is a complete, plausible-looking resume
 * with the wrong contents sitting in the directory you open when you
 * need to attach one. But a delete that is one careless regex edit
 * from matching too much is worse than the problem it solves, so the
 * bulk of this file is about what pruneStale must NOT touch: the
 * other document's outputs, the build's own JSON and HTML, a PDF you
 * put in dist/ yourself, and above all the files the current build
 * just wrote.
 */

const path = require('path');
const fs = require('fs');
const os = require('os');

const { assertEq, assertTrue, test, report } =
  require(path.join(__dirname, '_framework'));

const ROOT = path.join(__dirname, '..');
const on = require(path.join(ROOT, 'build', '_output_name.js'));


/** A throwaway directory, removed when `fn` returns. */
function withTempDir(fn) {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'outputname-'));
  try {
    return fn(dir);
  } finally {
    fs.rmSync(dir, { recursive: true, force: true });
  }
}

function writeMeta(dir, obj) {
  const p = path.join(dir, 'pdf_meta.json');
  fs.writeFileSync(p, typeof obj === 'string' ? obj : JSON.stringify(obj), 'utf-8');
  return p;
}

function touch(dir, ...names) {
  for (const n of names) fs.writeFileSync(path.join(dir, n), 'x', 'utf-8');
}

const listing = dir => fs.readdirSync(dir).sort();


/* ─── reading the stem back out of the metadata ───────────────── */

test('the stem comes from the build metadata, not from a guess', () => {
  withTempDir((dir) => {
    const meta = writeMeta(dir, { output_stem: 'Gaius_Caesar_Resume' });
    const p = on.outputPaths(dir, meta, 'resume');
    assertEq(p.stem, 'Gaius_Caesar_Resume', 'stem');
    assertEq(path.basename(p.colorPdf), 'Gaius_Caesar_Resume.pdf', 'color');
    assertEq(path.basename(p.grayscalePdf),
             'Gaius_Caesar_Resume_Grayscale.pdf', 'grayscale');
  });
});

test('the color variant takes the bare stem', () => {
  // The file you attach to an application gets the clean name; only
  // the black-and-white one is suffixed. This is the user's choice
  // and the reason the two paths are not symmetric.
  withTempDir((dir) => {
    const meta = writeMeta(dir, { output_stem: 'A_B_Cover_Letter' });
    const p = on.outputPaths(dir, meta, 'letter');
    assertEq(path.basename(p.colorPdf), 'A_B_Cover_Letter.pdf', 'no suffix');
    assertTrue(path.basename(p.grayscalePdf).endsWith(on.GRAYSCALE_SUFFIX + '.pdf'),
               'grayscale suffixed');
  });
});

test('every broken-metadata case falls back instead of throwing', () => {
  // These getters are hit on every status poll the app makes,
  // including before the first build and during one. Throwing here
  // would take out the tray, not just the name.
  withTempDir((dir) => {
    assertEq(on.readStem(path.join(dir, 'nope.json'), 'resume'),
             'Resume', 'missing file');
    assertEq(on.readStem(writeMeta(dir, '{ not json'), 'resume'),
             'Resume', 'corrupt file');
    assertEq(on.readStem(writeMeta(dir, { title: 'x' }), 'resume'),
             'Resume', 'field absent (an older build)');
    assertEq(on.readStem(writeMeta(dir, { output_stem: '' }), 'resume'),
             'Resume', 'empty field');
    assertEq(on.readStem(writeMeta(dir, { output_stem: 42 }), 'letter'),
             'Cover_Letter', 'wrong type');
  });
});

test('the fallback names a file that correctly does not exist', () => {
  // The bootstrap case is not an error to report — nothing has been
  // built, so "not built" is the honest answer, and it is what the
  // absent fallback path produces.
  withTempDir((dir) => {
    const p = on.outputPaths(dir, path.join(dir, 'pdf_meta.json'), 'resume');
    assertEq(path.basename(p.colorPdf), 'Resume.pdf', 'fallback name');
    assertTrue(!fs.existsSync(p.colorPdf), 'and it is absent');
  });
});

test('an unknown variant is refused rather than guessed at', () => {
  let threw = false;
  try { on.readStem('whatever.json', 'invoice'); } catch { threw = true; }
  assertTrue(threw, 'readStem rejects');

  threw = false;
  try { on.outputPattern('invoice'); } catch { threw = true; }
  assertTrue(threw, 'outputPattern rejects');
});


/* ─── the pattern: what counts as this project's output ───────── */

test('the pattern matches this document\'s outputs, named or not', () => {
  const re = on.outputPattern('resume');
  assertTrue(re.test('Gaius_Caesar_Resume.pdf'), 'named color');
  assertTrue(re.test('Gaius_Caesar_Resume_Grayscale.pdf'), 'named grayscale');
  assertTrue(re.test('Resume.pdf'), 'the no-name fallback');
  assertTrue(re.test('Anne_Marie_Smith_Jones_Resume.pdf'), 'many parts');
});

test('the pattern does not reach past its own document', () => {
  const resume = on.outputPattern('resume');
  assertTrue(!resume.test('Gaius_Caesar_Cover_Letter.pdf'), 'the other document');
  assertTrue(!resume.test('pdf_meta.json'), 'build metadata');
  assertTrue(!resume.test('index.html'), 'the rendered HTML');
  assertTrue(!resume.test('styles.css'), 'the stylesheet');
  assertTrue(!resume.test('Resume.pdf.bak'), 'not anchored at the end');
  assertTrue(!resume.test('old_Resume.pdf.txt'), 'wrong extension');
  assertTrue(!resume.test('Resumes.pdf'), 'a longer word');
  assertTrue(!resume.test('My_Resume_Draft.pdf'), 'suffix must be last');

  const letter = on.outputPattern('letter');
  assertTrue(letter.test('Gaius_Caesar_Cover_Letter_Grayscale.pdf'), 'its own');
  assertTrue(!letter.test('Gaius_Caesar_Resume.pdf'), 'the other document');
});


/* ─── pruning: the part that deletes ──────────────────────────── */

test('pruneStale removes the previous name and keeps the current one', () => {
  withTempDir((dir) => {
    touch(dir,
      'Gaius_Caesar_Resume.pdf',            // this build
      'Gaius_Caesar_Resume_Grayscale.pdf',  // this build
      'Gaius_Julius_Resume.pdf');           // yesterday's spelling

    const keep = [
      path.join(dir, 'Gaius_Caesar_Resume.pdf'),
      path.join(dir, 'Gaius_Caesar_Resume_Grayscale.pdf'),
    ];
    const removed = on.pruneStale(dir, 'resume', keep);

    assertEq(removed.join(','), 'Gaius_Julius_Resume.pdf', 'removed the stale one');
    assertEq(listing(dir).join(','),
             'Gaius_Caesar_Resume.pdf,Gaius_Caesar_Resume_Grayscale.pdf',
             'the current pair survives');
  });
});

test('pruneStale removes the pre-rename legacy filenames', () => {
  // The migration case: everyone has a resume-color.pdf from before
  // outputs were named after people.
  withTempDir((dir) => {
    touch(dir, 'resume-color.pdf', 'resume-grayscale.pdf', 'Gaius_Caesar_Resume.pdf');
    on.pruneStale(dir, 'resume', [path.join(dir, 'Gaius_Caesar_Resume.pdf')]);
    assertEq(listing(dir).join(','), 'Gaius_Caesar_Resume.pdf', 'only today\'s left');
  });
});

test('pruneStale removes a variant written under the retired suffix', () => {
  // The grayscale marker was respelled ('-grayscale' → '_Grayscale').
  // A file written the old way is not the file this build wrote and
  // is not overwritten by it, so if the pattern stopped matching it,
  // it would stay in dist/ looking current forever.
  withTempDir((dir) => {
    touch(dir, 'Gaius_Caesar_Resume.pdf',
               'Gaius_Caesar_Resume_Grayscale.pdf',
               'Gaius_Caesar_Resume-grayscale.pdf');   // yesterday's spelling
    const keep = ['Gaius_Caesar_Resume.pdf', 'Gaius_Caesar_Resume_Grayscale.pdf']
      .map(n => path.join(dir, n));
    const removed = on.pruneStale(dir, 'resume', keep);
    assertEq(removed.join(','), 'Gaius_Caesar_Resume-grayscale.pdf', 'the old one');
    assertEq(listing(dir).length, 2, 'the current pair survives');
  });
});

test('the retired suffixes widen the pattern and nothing else', () => {
  // They must never reach outputPaths — a build writes the current
  // spelling only.
  withTempDir((dir) => {
    const meta = writeMeta(dir, { output_stem: 'A_B_Resume' });
    assertEq(path.basename(on.outputPaths(dir, meta, 'resume').grayscalePdf),
             'A_B_Resume' + on.GRAYSCALE_SUFFIX + '.pdf', 'writes the current spelling');
  });
  const re = on.outputPattern('resume');
  for (const retired of on.RETIRED_GRAYSCALE_SUFFIXES) {
    assertTrue(re.test(`A_B_Resume${retired}.pdf`), `matches ${retired}`);
    assertTrue(retired !== on.GRAYSCALE_SUFFIX, `${retired} is actually retired`);
  }
});

test('pruneStale never touches the other document', () => {
  // Building a resume must not delete the cover letter you built an
  // hour ago. The two documents share dist/ and nothing else.
  withTempDir((dir) => {
    touch(dir,
      'Gaius_Caesar_Resume.pdf',
      'Gaius_Caesar_Cover_Letter.pdf',
      'Gaius_Caesar_Cover_Letter_Grayscale.pdf',
      'letter-color.pdf');
    on.pruneStale(dir, 'resume', [path.join(dir, 'Gaius_Caesar_Resume.pdf')]);
    assertEq(listing(dir).length, 4, 'nothing was removed');
  });
});

test('pruneStale never touches anything outside its own pattern', () => {
  withTempDir((dir) => {
    touch(dir,
      'Gaius_Caesar_Resume.pdf',
      'pdf_meta.json', 'placement.json', 'index.html', 'styles.css',
      'favicon.svg',
      'Some_Other_Document.pdf',   // a PDF the user dropped in dist/
      'notes.txt');
    on.pruneStale(dir, 'resume', [path.join(dir, 'Gaius_Caesar_Resume.pdf')]);
    assertEq(listing(dir).length, 8, 'everything survives');
  });
});

test('a variant this build chose not to make is removed, not orphaned', () => {
  // RESUME_VARIANTS=color leaves the grayscale file out of `keep`, so
  // the previous run's grayscale PDF must go. Otherwise "the file
  // exists" stops meaning "this build made it" — which the snapshot
  // test and the app both rely on.
  withTempDir((dir) => {
    touch(dir, 'Gaius_Caesar_Resume.pdf', 'Gaius_Caesar_Resume_Grayscale.pdf');
    on.pruneStale(dir, 'resume', [path.join(dir, 'Gaius_Caesar_Resume.pdf')]);
    assertEq(listing(dir).join(','), 'Gaius_Caesar_Resume.pdf', 'grayscale gone');
  });
});

test('pruneStale compares resolved paths, not the strings it was given', () => {
  withTempDir((dir) => {
    touch(dir, 'Gaius_Caesar_Resume.pdf');
    // The same file, spelled the long way round.
    const awkward = path.join(dir, 'sub', '..', 'Gaius_Caesar_Resume.pdf');
    fs.mkdirSync(path.join(dir, 'sub'));
    const removed = on.pruneStale(dir, 'resume', [awkward]);
    assertEq(removed.length, 0, 'recognized as the kept file');
    assertTrue(fs.existsSync(path.join(dir, 'Gaius_Caesar_Resume.pdf')), 'survived');
  });
});

/* ─── case-insensitive filesystems ─────────────────────────────
 *
 * On NTFS / APFS, writing Gaius_CAESAR_Resume.pdf over an existing
 * Gaius_Caesar_Resume.pdf keeps the old spelling on disk. The keep-list
 * then names a path that readdir never returns, and a string compare
 * would delete the PDF the build just wrote. */

test('pruneStale keeps a file that IS a kept path under another spelling', () => {
  // Simulated on any OS with a hard link: dist/ holds the old-case name
  // only; the keep-list names the new-case path, which (as on a
  // case-insensitive volume) is the same file — same dev and inode.
  withTempDir((dir) => {
    const dist = path.join(dir, 'dist');
    const elsewhere = path.join(dir, 'other');
    fs.mkdirSync(dist);
    fs.mkdirSync(elsewhere);
    touch(dist, 'Gaius_Caesar_Resume.pdf');
    const newCase = path.join(elsewhere, 'Gaius_CAESAR_Resume.pdf');
    try {
      fs.linkSync(path.join(dist, 'Gaius_Caesar_Resume.pdf'), newCase);
    } catch (err) {
      // No hard links on this filesystem: the stubbed test below still
      // covers the logic.
      assertTrue(true, `hard links unavailable (${err.code}) — covered by the stub test`);
      return;
    }
    const removed = on.pruneStale(dist, 'resume', [newCase]);
    assertEq(removed.length, 0, 'the same file is recognized by identity');
    assertTrue(fs.existsSync(path.join(dist, 'Gaius_Caesar_Resume.pdf')),
               'the PDF the build just wrote survives');
  });
});

/** A one-file case-insensitive fs with no inodes (ino 0), like some network shares. */
function caseInsensitiveFs(dist, onDiskName) {
  const onDisk = path.join(dist, onDiskName).toLowerCase();
  const deleted = [];
  const exists = p => path.resolve(p).toLowerCase() === onDisk && !deleted.length;
  const enoent = (p) => Object.assign(new Error(`ENOENT: ${p}`), { code: 'ENOENT' });
  return {
    deleted,
    fs: {
      readdirSync: d => (path.resolve(d) === path.resolve(dist) && !deleted.length
        ? [onDiskName] : []),
      statSync: (p, opts) => {
        if (!exists(p)) throw enoent(p);
        const zero = opts && opts.bigint ? 0n : 0;
        return { dev: zero, ino: zero, isFile: () => true };
      },
      rmSync: (p) => { deleted.push(path.basename(p)); },
    },
  };
}

test('with no inode, win32/darwin fall back to a case-insensitive compare', () => {
  const dist = path.resolve('/virtual/dist');
  for (const platform of ['win32', 'darwin']) {
    const stub = caseInsensitiveFs(dist, 'Gaius_Caesar_Resume.pdf');
    const removed = on.pruneStale(dist, 'resume',
      [path.join(dist, 'Gaius_CAESAR_Resume.pdf')], null, null,
      { fs: stub.fs, platform });
    assertEq(removed, [], `${platform}: nothing removed`);
    assertEq(stub.deleted, [], `${platform}: the just-written PDF is not deleted`);
  }
});

test('with no inode on a case-sensitive platform, a differently-cased name is stale', () => {
  // Linux: Gaius_CAESAR_Resume.pdf and Gaius_Caesar_Resume.pdf are two
  // files, and the old one really is stale. ino 0 must not make every
  // file look identical (dev:0 / ino:0 is "unknown", not a match).
  const dist = path.resolve('/virtual/dist');
  const stub = caseInsensitiveFs(dist, 'Gaius_Caesar_Resume.pdf');
  const removed = on.pruneStale(dist, 'resume',
    [path.join(dist, 'Gaius_CAESAR_Resume.pdf')], null, null,
    { fs: stub.fs, platform: 'linux' });
  assertEq(removed, ['Gaius_Caesar_Resume.pdf'], 'the old spelling is pruned');
});

test('identity never spares an unrelated stale file', () => {
  withTempDir((dir) => {
    touch(dir, 'Gaius_Caesar_Resume.pdf', 'Gaius_Julius_Resume.pdf');
    const removed = on.pruneStale(dir, 'resume', [path.join(dir, 'Gaius_Caesar_Resume.pdf')]);
    assertEq(removed, ['Gaius_Julius_Resume.pdf'], 'a different file is still removed');
  });
});

test('a directory sharing the naming pattern is not deleted', () => {
  withTempDir((dir) => {
    fs.mkdirSync(path.join(dir, 'Gaius_Caesar_Resume.pdf'));
    on.pruneStale(dir, 'resume', []);
    assertTrue(fs.existsSync(path.join(dir, 'Gaius_Caesar_Resume.pdf')),
               'directories are skipped');
  });
});

test('a missing dist/ is not an error', () => {
  withTempDir((dir) => {
    const removed = on.pruneStale(path.join(dir, 'no-such-dir'), 'resume', []);
    assertEq(removed.length, 0, 'returns empty rather than throwing');
  });
});

test('every removal is reported, so a build says what it deleted', () => {
  // Silent deletion of a file that looks like your resume is not
  // acceptable even when it is the right call.
  withTempDir((dir) => {
    touch(dir, 'Old_Name_Resume.pdf', 'resume-color.pdf');
    const seen = [];
    const removed = on.pruneStale(dir, 'resume', [], name => seen.push(name));
    assertEq(seen.sort().join(','), 'Old_Name_Resume.pdf,resume-color.pdf',
             'both announced');
    assertEq(seen.length, removed.length, 'one callback per removal');
  });
});


/* ─── against the real project ────────────────────────────────── */

test('the constants the two languages share are present and sane', () => {
  assertEq(on.DOC_SUFFIX.resume, 'Resume', 'resume suffix');
  assertEq(on.DOC_SUFFIX.letter, 'Cover_Letter', 'letter suffix');
  assertEq(on.SEPARATOR, '_', 'separator');
  assertEq(on.GRAYSCALE_SUFFIX, '_Grayscale', 'grayscale suffix');
  // The grayscale marker shares the separator with the name parts.
  // That is safe only because DOC_SUFFIX always ends the stem, so the
  // marker can only ever appear after it — never inside a name.
  assertTrue(!on.DOC_SUFFIX.resume.includes(on.GRAYSCALE_SUFFIX),
             'the suffixes do not overlap');
  assertTrue(!on.DOC_SUFFIX.letter.includes(on.GRAYSCALE_SUFFIX),
             'nor for the letter');
  assertTrue(!on.RETIRED_GRAYSCALE_SUFFIXES.includes(on.GRAYSCALE_SUFFIX),
             'the current spelling is not listed as retired');
});

test('the pipeline hands back paths that match this module', () => {
  // The integration point: pipeline.paths.colorPdf is a getter over
  // outputPaths, and resume.js prunes with the result. If these two
  // ever disagreed, a build would delete the file it just wrote.
  const { createPipeline } = require(path.join(ROOT, 'build', 'pipeline.js'));
  const noPython = { buildHtml() {}, cropPdf() {} };
  for (const variant of ['resume', 'letter']) {
    const p = createPipeline({ root: ROOT, python: noPython, variant }).paths;
    const expected = on.outputPaths(path.join(ROOT, 'dist'), p.pdfMeta, variant);
    assertEq(p.colorPdf, expected.colorPdf, `${variant} color`);
    assertEq(p.grayscalePdf, expected.grayscalePdf, `${variant} grayscale`);
    assertTrue(on.outputPattern(variant).test(path.basename(p.colorPdf)),
               `${variant} color matches the prune pattern`);
    assertTrue(on.outputPattern(variant).test(path.basename(p.grayscalePdf)),
               `${variant} grayscale matches the prune pattern`);
  }
});

report('output_name');
